src/dmd/declaration.d

```diff
@@ -1454,14 +1454,24 @@ extern (C++) class VarDeclaration : Declaration
                 //if (cd.isInterfaceDeclaration())
                 //    error("interface `%s` cannot be scope", cd.toChars());
 
-                // Destroying C++ scope classes crashes currently. Since C++ class dtors are not currently supported, simply do not run dtors for them.
-                // See https://issues.dlang.org/show_bug.cgi?id=13182
-                if (cd.classKind == ClassKind.cpp)
-                {
-                    break;
-                }
                 if (mynew || onstack) // if any destructors
                 {
+                    // delete'ing C++ classes crashes (and delete is deprecated anyway)
+                    if (cd.classKind == ClassKind.cpp)
```

Comment:
`destroy` doesn't work for `const`/`immutable` either (unless we insert an explicit cast to mutable)

Reply:
Implemented another approach, it's now directly calling the dtor (effectively inlining the call to `destroy`).
This matches the implementation for `struct`s and allows the *hack* to call dtors on immutable instances.

Reply:
> destroy doesn't work for const/immutable either (unless we insert an explicit cast to mutable)

Seems like it does (tested with v2.096.0). Anyway, the direct dtor call is better. Is there a reason you don't use it for D classes too? I find it inconsistent (can use `scope const/immutable` with C++ classes, but not with D ones) and don't really like deprecated expressions in compiler-generated AST nodes. ;)

Reply:
> Seems like it does (tested with v2.096.0).

Nevermind, seems like it was an issue with the lowering.

> Is there a reason you don't use it for D classes too? I find it inconsistent (can use scope const/immutable with C++ classes, but not with D ones) and don't really like deprecated expressions in compiler-generated AST nodes. ;)

Mostly because `destroy` (and `delete` IIRC) refer to `rt_finalize2` which does more cleanup than just calling the dtor. Not sure if those additional steps are required for stack allocated instances...

But lowering to `destroy` would still work for D classes.

Reply:
> But lowering to destroy would still work for D classes.

Yeah that'd be great.

You're right, for D classes, there's apparently the need to descend to all base classes and invoke the dtors manually (via TypeInfo... argh), and the monitor might need cleanup too. LDC has a special case for these `delete` expressions for `scope` classes and tries to avoid calling druntime if the class has no finalizer *and* the monitor is null: https://github.com/ldc-developers/ldc/blob/937c57b8354d4821daededb35abdb77d16ce3737/gen/classes.cpp#L200-L222

Doing this here in the frontend directly would obviously be much nicer.

Reply:
Maybe the best way forward is to replace the `TypeInfo` based dtor-loop in DRuntime with a real dtor chain as done for C++ classes and to move the cleanup into `Object.~this()). That would cause less (magic) code scattered across dmd/druntime/ldc/... .

Anyway, lowering to `destroy` doesn't work yet.
- it isn't alway `@nogc`
- CTFE cannot handle `destroy` (it needs some special casing as done for `delete`)